### PR TITLE
Convert bins to numeric for histogram if a string

### LINKED ISSUE #29484: DataFrame histogram plot error with bins='auto'
Hi,

Im having a problem using pandas.plot.hist with the bins argument set to 'auto'.
I'm on os x, with the following package versions installed:

pandas: '0.25.3'
numpy: '1.17.2'
matplotlib: '3.1.1'

This error happens if I'm in jupyterlab, notebook, python or ipython.

#### Here's what produces the error:

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

test_data = np.random.randint(0, 100, 10000)
dataset = pd.DataFrame({'test': test_data})

dataset.plot.hist(bins='auto')
```

#### Output of the above code:

<details>

ValueError                                Traceback (most recent call last)
<ipython-input-16-520a066d2367> in <module>
      6 dataset = pd.DataFrame({'test': test_data})
      7
----> 8 dataset.plot.hist(bins='auto')

~/anaconda3/envs/default/lib/python3.7/site-packages/pandas/plotting/_core.py in hist(self, by, bins, **kwargs)
   1123             >>> ax = df.plot.hist(bins=12, alpha=0.5)
   1124         """
-> 1125         return self(kind="hist", by=by, bins=bins, **kwargs)
   1126
   1127     def kde(self, bw_method=None, ind=None, **kwargs):

~/anaconda3/envs/default/lib/python3.7/site-packages/pandas/plotting/_core.py in __call__(self, *args, **kwargs)
    792                     data.columns = label_name
    793
--> 794         return plot_backend.plot(data, kind=kind, **kwargs)
    795
    796     def line(self, x=None, y=None, **kwargs):

~/anaconda3/envs/default/lib/python3.7/site-packages/pandas/plotting/_matplotlib/__init__.py in plot(data, kind, **kwargs)
     60             kwargs["ax"] = getattr(ax, "left_ax", ax)
     61     plot_obj = PLOT_CLASSES[kind](data, **kwargs)
---> 62     plot_obj.generate()
     63     plot_obj.draw()
     64     return plot_obj.result

~/anaconda3/envs/default/lib/python3.7/site-packages/pandas/plotting/_matplotlib/core.py in generate(self)
    279         self._compute_plot_data()
    280         self._setup_subplots()
--> 281         self._make_plot()
    282         self._add_table()
    283         self._make_legend()

~/anaconda3/envs/default/lib/python3.7/site-packages/pandas/plotting/_matplotlib/hist.py in _make_plot(self)
     81
     82             kwds = self._make_plot_keywords(kwds, y)
---> 83             artists = self._plot(ax, y, column_num=i, stacking_id=stacking_id, **kwds)
     84             self._add_legend_handle(artists[0], label, index=i)
     85

~/anaconda3/envs/default/lib/python3.7/site-packages/pandas/plotting/_matplotlib/hist.py in _plot(cls, ax, y, style, bins, bottom, column_num, stacking_id, **kwds)
     60         bottom = bottom + cls._get_stacked_values(ax, stacking_id, base, kwds["label"])
     61         # ignore style
---> 62         n, bins, patches = ax.hist(y, bins=bins, bottom=bottom, **kwds)
     63         cls._update_stacker(ax, stacking_id, n)
     64         return patches

~/anaconda3/envs/default/lib/python3.7/site-packages/matplotlib/__init__.py in inner(ax, data, *args, **kwargs)
   1599     def inner(ax, *args, data=None, **kwargs):
   1600         if data is None:
-> 1601             return func(ax, *map(sanitize_sequence, args), **kwargs)
   1602
   1603         bound = new_sig.bind(ax, *args, **kwargs)

~/anaconda3/envs/default/lib/python3.7/site-packages/matplotlib/axes/_axes.py in hist(self, x, bins, range, density, weights, cumulative, bottom, histtype, align, orientation, rwidth, log, color, label, stacked, normed, **kwargs)
   6841                 patch = _barfunc(bins[:-1]+boffset, height, width,
   6842                                  align='center', log=log,
-> 6843                                  color=c, **{bottom_kwarg: bottom})
   6844                 patches.append(patch)
   6845                 if stacked:

~/anaconda3/envs/default/lib/python3.7/site-packages/matplotlib/__init__.py in inner(ax, data, *args, **kwargs)
   1599     def inner(ax, *args, data=None, **kwargs):
   1600         if data is None:
-> 1601             return func(ax, *map(sanitize_sequence, args), **kwargs)
   1602
   1603         bound = new_sig.bind(ax, *args, **kwargs)

~/anaconda3/envs/default/lib/python3.7/site-packages/matplotlib/axes/_axes.py in bar(self, x, height, width, bottom, align, **kwargs)
   2373         x, height, width, y, linewidth = np.broadcast_arrays(
   2374             # Make args iterable too.
-> 2375             np.atleast_1d(x), height, width, y, linewidth)
   2376
   2377         # Now that units have been converted, set the tick locations.

<__array_function__ internals> in broadcast_arrays(*args, **kwargs)

~/anaconda3/envs/default/lib/python3.7/site-packages/numpy/lib/stride_tricks.py in broadcast_arrays(*args, **kwargs)
    262     args = [np.array(_m, copy=False, subok=subok) for _m in args]
    263
--> 264     shape = _broadcast_shape(*args)
    265
    266     if all(array.shape == shape for array in args):

~/anaconda3/envs/default/lib/python3.7/site-packages/numpy/lib/stride_tricks.py in _broadcast_shape(*args)
    189     # use the old-iterator because np.nditer does not handle size 0 arrays
    190     # consistently
--> 191     b = np.broadcast(*args[:32])
    192     # unfortunately, it cannot handle 32 or more arguments directly
    193     for pos in range(32, len(args), 31):

ValueError: shape mismatch: objects cannot be broadcast to a single shape

</details>

Thanks for any help with this. I think it may be specific to os x, I didn't used to see this error in windows/linux (although I don't have access to test this out on other systems currently to confirm this)

### PR DESCRIPTION
**Pull request by wesbarnett**
Fixes #29484

- [ ] closes #29484
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by WillAyd — CHANGES_REQUESTED** [review]
Can you add test(s)? We typically look for these as the first part of any PR

**Comment by wesbarnett** [comment]
Yes I'll work on creating a test.

**Comment by jreback** [comment]
can you merge master and add a test

**Comment by WillAyd** [comment]
@wesbarnett does this type of bins handling come from other packages? We only have this documented as working for `integer or sequence`

**Comment by wesbarnett** [comment]
@WillAyd it comes from matplotlib where you can pass things like "auto". See #29484.

**Comment by WillAyd** [comment]
Makes sense. Can you update the docstring as well then?

**Comment by wesbarnett** [comment]
@WillAyd I only see the docstring for hist indicating an integer is supported. I've added a blurb taken from the matplotlib docstring to indicate that a string is supported

Is sequence also supported?

File: pandas/plotting/_core.py [hunk 1]
```diff
@@ -1178,8 +1178,12 @@ def hist(self, by=None, bins=10, **kwargs):
         ----------
         by : str or sequence, optional
             Column in the DataFrame to group by.
-        bins : int, default 10
+        bins : int, default 10 or string
```
Comment (WillAyd) [inline pandas/plotting/_core.py]:
```suggestion
        bins : str or int, default 10
```

File: pandas/tests/plotting/test_hist_method.py [hunk 2]
```diff
@@ -169,6 +169,10 @@ def test_hist_df_legacy(self):
         with tm.assert_produces_warning(UserWarning):
             _check_plot_works(df.hist, bins=5)
 
+        # check bins argument with string
+        with tm.assert_produces_warning(UserWarning):
+            _check_plot_works(df.hist, bins="auto")
```
Comment (WillAyd) [inline pandas/tests/plotting/test_hist_method.py]: Can you write a test that actually ensures the binning behavior of "auto", i.e. not just that this works without error but that it also actually provides back the correct bin values?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/plotting/test_hist_method.py (hunk 2 above)
Comment (wesbarnett) [reply pandas/tests/plotting/test_hist_method.py]: So you want me to check that `np.histogram` works? Should I also check every possible valid string that `np.histogram` takes?

Comment (WillAyd) [reply pandas/tests/plotting/test_hist_method.py]: Not asking to check `np.histogram` but rather that the plot renders the ticklabels of the rendered plot are as expected. I think can just use `get_xticklabels` to be sure

Comment (wesbarnett) [reply pandas/tests/plotting/test_hist_method.py]: Honestly not sure how to go about this other than manually creating a couple of histograms and just checking the locations of the bins. The xtick labels could be the same for a histogram that uses "auto" and one that doesn't. For example:

```
df = DataFrame(randn(1000))
df.hist(bins="auto")
```
image
```
df.hist()
```
image

Comment (WillAyd) [reply pandas/tests/plotting/test_hist_method.py]: I think that's fine; even just checking for those labels would improve the test, knowing they don't stray from the default

File: pandas/plotting/_matplotlib/hist.py [hunk 3]
```diff
@@ -49,14 +49,25 @@ def _plot(
...
+
         if column_num == 0:
             cls._initialize_stacker(ax, stacking_id, len(bins) - 1)
-        y = y[~isna(y)]
 
         base = np.zeros(len(bins) - 1)
         bottom = bottom + cls._get_stacked_values(ax, stacking_id, base, kwds["label"])
         # ignore style
-        n, bins, patches = ax.hist(y, bins=bins, bottom=bottom, **kwds)
+        n, _, patches = ax.hist(y, bins=bins, bottom=bottom, **kwds)
```
Comment (WillAyd) [inline pandas/plotting/_matplotlib/hist.py]: Is there a particular reason that we have a separate call to `np.histogram` above to get the bins? Is it not possible to return the appropriate values from the `ax.hist` call here?

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
@wesbarnett can you take a look at last round of comments?

File: pandas/plotting/_matplotlib/hist.py (hunk 3 above)
Comment (wesbarnett) [reply pandas/plotting/_matplotlib/hist.py:70]: `cls._initialize_stacker` needs the number of bins. `ax.hist` can't provide that yet, since it needs `bottom`, which comes from `cls._get_stacked_values` and that can't be called before `cls._initialize_stacker`. That's my understanding.

**Comment by jreback** [comment]
@wesbarnett this also needs a whatnew note (bug fixes plotting in 1.1), pls ping on green when ready.

**Comment by wesbarnett** [comment]
> @wesbarnett can you take a look at last round of comments?

See the previous thread. I think I addressed everything. Sorry if I missed anything.

**Comment by wesbarnett** [comment]
> @wesbarnett this also needs a whatnew note (bug fixes plotting in 1.1), pls ping on green when ready.

@jreback  I'm having trouble locating where to add this (I've looked in README and RELEASES). I don't see a changelog where this might go. This is my first PR for pandas.

**Comment by jreback** [comment]
@wesbarnett can you merge master

**Comment by TomAugspurger** [comment]
Merged master to fix the upstream CI issue.

@wesbarnett do we have a test for this that uses `.plot.hist()`?

**Comment by wesbarnett** [comment]
@TomAugspurger No test for `.plot.hist()` in this. I can add some.

**Review by jreback — CHANGES_REQUESTED** [review]
can you merge master, you are picking up other commits.

**Comment by wesbarnett** [comment]
@jreback Okay, should be good now. I merged master again.

File: pandas/plotting/_core.py [hunk 4]
```diff
@@ -1213,8 +1213,12 @@ def hist(self, by=None, bins=10, **kwargs):
         ----------
         by : str or sequence, optional
             Column in the DataFrame to group by.
-        bins : int, default 10
+        bins : str or int, default 10
             Number of histogram bins to be used.
+
+            You can alternatively provide a string describing a binning
+            strategy, such as 'auto', 'sturges', 'fd', 'doane', 'scott',
+            'rice' or 'sqrt'. See `numpy.histogram`.
```
Comment (TomAugspurger) [inline pandas/plotting/_core.py:1221]:
```suggestion
            'rice' or 'sqrt'. See :func:`numpy.histogram`.
```

File: pandas/plotting/_matplotlib/hist.py [hunk 5]
```diff
@@ -44,14 +44,25 @@ def _plot(
...
+
+        # if bins is a string (e.g. "auto"), get actual bins
+        # so that it can be used in creating stackers below
+        _, bins = np.histogram(
+            y,
+            bins=bins,
+            range=kwds.get("range", None),
+            weights=kwds.get("weights", None),
+        )
+
```
Comment (TomAugspurger) [inline pandas/plotting/_matplotlib/hist.py:58]: Why is this necessary? Why can we just pass the string to matplotlib?

File: doc/source/whatsnew/v1.1.0.rst [hunk 6]
```diff
@@ -1005,6 +1005,7 @@ Plotting
 - Bug in :meth:`DataFrame.boxplot` and :meth:`DataFrame.plot.boxplot` lost color attributes of ``medianprops``, ``whiskerprops``, ``capprops`` and ``medianprops`` (:issue:`30346`)
 - Bug in :meth:`DataFrame.hist` where the order of ``column`` argument was ignored (:issue:`29235`)
 - Bug in :meth:`DataFrame.plot.scatter` that when adding multiple plots with different ``cmap``, colorbars alway use the first ``cmap`` (:issue:`33389`)
+- Bug in :meth: `DataFrame.plot` for plotting histograms when passing a string value such as `auto` (:issue:`29484`).
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.1.0.rst:1008]:
```suggestion
- Bug in :meth:`DataFrame.plot` for plotting histograms when passing a string value such as ``'auto'`` (:issue:`29484`).
```

File: pandas/plotting/_core.py [hunk 7]
```diff
@@ -1213,8 +1213,12 @@ def hist(self, by=None, bins=10, **kwargs):
         ----------
         by : str or sequence, optional
             Column in the DataFrame to group by.
-        bins : int, default 10
+        bins : str or int, default 10
```
Comment (TomAugspurger) [inline pandas/plotting/_core.py:1216]:
```suggestion
        bins : int or str, default 10
```

the default typically goes first.

File: pandas/plotting/_matplotlib/hist.py (hunk 5 above)
Comment (wesbarnett) [reply pandas/plotting/_matplotlib/hist.py:58]: The comment above it explains why - we need the value of bins to be numeric because it is used in creating stackers here: https://github.com/pandas-dev/pandas/blob/129b62fbe8e2aa153ecdaccdcc14a8846e430060/pandas/plotting/_matplotlib/hist.py#L59-L62

The results from initializing the stacker are required to create the plot (specifically the argument `bottom`).

Comment (TomAugspurger) [reply pandas/plotting/_matplotlib/hist.py:58]: Matplotlib returns `bins` from `hist`. Can we pass it the string capture that?

**Comment by TomAugspurger** [comment]
@wesbarnett merging master should fix the CI failures.

File: pandas/plotting/_matplotlib/hist.py (hunk 5 above)
Comment (wesbarnett) [reply pandas/plotting/_matplotlib/hist.py:58]: You're correct that it returns bins (2nd arg here):

https://github.com/pandas-dev/pandas/blob/129b62fbe8e2aa153ecdaccdcc14a8846e430060/pandas/plotting/_matplotlib/hist.py#L65

But that needs `bottom`. `bottom` needs `base` and `base` needs `bins` to be a list not a string (see lines above).

**Comment by WillAyd** [comment]
@wesbarnett is this still active? Can you move note to 1.2?

**Comment by wesbarnett** [comment]
@WillAyd @jreback this is ready for review. One of the builds failed to run. Not sure how to retrigger that.

**Comment by arw2019** [comment]
ping @WillAyd @jreback 

@wesbarnett can you merge master again?

**Comment by jreback** [comment]
closing as stale. this looks like an ok soln, though likely needs some more tests.
